### ISSUE TITLE
为什么catch不到DioError ，总是跳到throw assureDioError(e, options);

### ISSUE THREAD

**Issue by wish32** [leancloud/Storage-SDK-Flutter#98]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | e.g. flutter / windows10/ android                          |      |
| Platform Version               | e.g.  1.22.4 / 10 / 9.0                              |      |
| Dio Version                    | e.g.  3.0.10                                   |      |
| Visual Studio Code Version | e.g.   1.52.1              |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |      |

### Issue Description and Steps

我使用leancloud的SDK，当对方服务器返回400错误的时候，vsc 就会跳到这个页面：

dio.dart 第966行
} catch (e) {
      throw assureDioError(e, options);
}
并有红色提示：
Exception has occurred.
DioError (DioError [DioErrorType.RESPONSE]: Http status error [400])

于是我跟踪到此SDK的post动作，输出一些发送的数据与参数：
SDK 的文件 lc_http_client.dart , line70:

try {
      print('=====1.The params to post:====={');
      print(path);
      print(options);
      print(data);
      print(queryParams);
      print('=====1.The params to post:=====}');

      Response response = await _dio.post(path,
          options: options, data: data, queryParameters: queryParams);
      return response.data;
    } on DioError catch (e) {
      print('=====2.Here catch DioError:====={');
      print(e);
      print('=====2.Here catch DioError:=====}');
      _handleError(e);  //此函数可以在这里点击查看 [ _handleError(e);](https://github.com/leancloud/Storage-SDK-Flutter/blob/985f9ad91e3b1d5824fd3340c68d6010a6c606e3/lib/internal/http/lc_http_client.dart)
    }

我也附上我本身的业务逻辑的代码：
try {
      developer.log('======Begin :=======');
      await LCSMSClient.requestSMSCode(_phoneNumber);
    }on DioError catch (e){
      developer.log('======DioError catch=======');
    } on LCException catch (e) {
      developer.log('======LCException catch=======');
      developer.log(e.message);
    } catch (e) {
      developer.log('======catch=======');
    }

输出如下：
[log] ======Begin :=======
I/flutter ( 1793): =====1.The params to post:====={
I/flutter ( 1793): requestSmsCode
I/flutter ( 1793): Instance of 'Options'
I/flutter ( 1793): {"mobilePhoneNumber":"+8625786"}
I/flutter ( 1793): null
I/flutter ( 1793): =====1.The params to post:=====}
I/flutter ( 1793): =====2.Here catch DioError:====={
I/flutter ( 1793): DioError [DioErrorType.RESPONSE]: Http status error [400]
I/flutter ( 1793): =====2.Here catch DioError:=====}
[log] ======LCException catch=======
[log] 无效的手机号码。

问题：看输出，dio.post执行之后是可以catch到DioError的，但是为什么VSC编辑器每次都会跳转到
throw assureDioError(e, options);
并有红色提示呢？

即使我改为如下，不用on DioError或改为catchError，还是会跳转到throw assureDioError(e, options);并有红色提示。

try {
    developer.log('======Begin :=======');
    await LCSMSClient.requestSMSCode(_phoneNumber).catchError(( /DioError/ e) {
        developer.log('======I am catchError=======');
    });
}catch (e) {
    developer.log('======catch=======');
}
输出：
V/AudioManager( 7109): querySoundEffectsEnabled...
[log] ======Begin :=======
[log] ======I am catchError=======

这里明显可以catch到，但为什么还会在throw assureDioError(e, options);里面抛出来呢？求大侠指导！

**Comment by onerain88** [comment]
是不是因为这个异常更底层，而 VSCode 开了 Exception Breakings 导致的断点呢？

**Comment by wish32** [comment]
你好，我使用了本库一下地址的测试例子还是会出现一样的状况，请问你测试会否也会呢？
https://github.com/leancloud/Storage-SDK-Flutter/blob/master/test/user_test.dart

整个文件简化为如下：

import 'package:flutter_test/flutter_test.dart';
import 'package:leancloud_storage/leancloud.dart';

import 'utils.dart';

void main() {
  group('user', () {
    setUp(() => initNorthChina());

    test('sign up', () async {
      LCUser user = new LCUser();
      user.username = DateTime.now().millisecondsSinceEpoch.toString();
      user.password = 'world';
      String email = '${DateTime.now().millisecondsSinceEpoch}@qq.com';
      user.email = email;
      String mobile = '26211';//此处故意改为无效的手机号码
      user.mobile = mobile;

      try {
        await user.signUp();
      }catch (e) {
        print('======catch you=======');
      }
    });

  });
}


调试信息：

Restarted application in 1,686ms.
I/flutter (19945): 00:00 +0: user sign up
I/flutter (19945): *** Request ***
I/flutter (19945): uri: https://ikggdre2.lc-cn-n1-shared.com/1.1/classes/_User
I/flutter (19945): method: POST
I/flutter (19945): responseType: ResponseType.json
I/flutter (19945): followRedirects: true
I/flutter (19945): connectTimeout: 0
I/flutter (19945): receiveTimeout: 0
I/flutter (19945): extra: {}
I/flutter (19945): headers:
I/flutter (19945):  X-LC-Id: ikGGdRE2YcVOemAaRbgp1xGJ-gzGzoHsz
I/flutter (19945):  User-Agent: LeanCloud-Flutter-SDK/0.6.0
I/flutter (19945):  content-type: application/json
I/flutter (19945):  X-LC-Sign: 8f5c9f95b97ef5ed2a9b8586dd1f1d94,1612077304194
I/flutter (19945): data:
I/flutter (19945): {username: 1612077304183, password: world, email: 1612077304184@qq.com, mobilePhoneNumber: 26211}
I/flutter (19945):
I/flutter (19945): *** DioError ***:
I/flutter (19945): uri: https://ikggdre2.lc-cn-n1-shared.com/1.1/classes/_User
I/flutter (19945): DioError [DioErrorType.RESPONSE]: Http status error [400]
I/flutter (19945): uri: https://ikggdre2.lc-cn-n1-shared.com/1.1/classes/_User
I/flutter (19945): statusCode: 400
I/flutter (19945): headers:
I/flutter (19945):  connection: keep-alive
I/flutter (19945):  cache-control: no-cache,no-store
I/flutter (19945):  date: Sun, 31 Jan 2021 07:15:05 GMT
I/flutter (19945):  transfer-encoding: chunked
I/flutter (19945):  content-type: application/json;charset=utf-8
I/flutter (19945):  pragma: no-cache
I/flutter (19945):  server: openresty
I/flutter (19945): Response Text:
I/flutter (19945): {"code":127,"error":"无效的手机号码。"}
I/flutter (19945):
I/flutter (19945): ======catch you=======
I/flutter (19945): 00:10 +1: All tests passed!
W/InputMethodManager(19945): startInputReason = 8
E/RtgSchedManager(19945): endActivityTransaction: margin state not match
W/libEGL  (19945): EGLNativeWindowType 0x7d0ecba010 disconnect failed
D/ZrHung.AppEyeUiProbe(19945): not watching, wait.
W/libEGL  (19945): EGLNativeWindowType 0x7d327f3cd0 disconnect failed
I/ActivityThread(19945): Handle window ActivityRecord{e5a650b token=android.os.BinderProxy@e185896 {ltd.cloudsoft.frankap/ltd.cloudsoft.frankap.MainActivity}} visibility: false
E/RtgSchedManager(19945): endActivityTransaction: margin state not match
D/DecorView(19945): showOrHideHighlightView: hasFocus=false; winMode=1; isMrgNull=true
D/RtgSched(19945): resetRtgSchedHandle failed enable:0
D/AVStandardWebSocketClient(19945): onWebsocketPong()
I/AwareBitmapCacher(19945): handleReleaseCache: pid=19945
I/AwareBitmapCacher(19945): init lrucache size: 2097152 pid=19945
W/AVStandardWebSocketClient(19945): onError socket=cn.leancloud.websocket.AVStandardWebSocketClient@b82f968, exception=Read error: ssl=0x7d87f2e588: I/O error during system call, Software caused connection abort
D/AVConnectionManager(19945): AVConnectionManager onError. client:cn.leancloud.websocket.AVStandardWebSocketClient@b82f968, exception:Read error: ssl=0x7d87f2e588: I/O error during system call, Software caused connection abort
I/PushService(19945): PushService#onDestroy
D/AVStandardWebSocketClient(19945): onClose socket=cn.leancloud.websocket.AVStandardWebSocketClient@b82f968, code=1006, message=Connectivity broken
D/AVStandardWebSocketClient(19945): send ping packet
W/AVStandardWebSocketClient(19945): onError socket=cn.leancloud.websocket.AVStandardWebSocketClient@b82f968, exception=Write error: ssl=0x7d87f2e588: I/O error during system call, Broken pipe
D/AVConnectionManager(19945): AVConnectionManager onError. client:cn.leancloud.websocket.AVStandardWebSocketClient@b82f968, exception:Write error: ssl=0x7d87f2e588: I/O error during system call, Broken pipe
D/AVConnectionManager(19945): client(cn.leancloud.websocket.AVStandardWebSocketClient@b82f968) closed...
D/AVConnectionManager(19945): reConnect rtm server. count=0
D/AppRouter(19945): fetchRTMServerFromRemote. router=https://bniipgp0.lc-cn-e1-shared.com, appId=BNiIPgp0xCWH8Cg5ddYP77bS-9Nh9j0Va, installationId=a680b25e69f8248a5449ad02639c92bd


仍旧还是在dio.dart ,第966行停止了，并提示：
Exception has occurred.
DioError (DioError [DioErrorType.RESPONSE]: Http status error [400])

**Comment by onerain88** [comment]
我用你的测试是正常的，请检查一下你的断点面板

![image](https://user-images.githubusercontent.com/1450714/106411315-f3de7a00-647f-11eb-88a4-4238c6b3cee9.png)

**Comment by wish32** [comment]
@onerain88 
谢谢，原来是VS code可以设置的，在左侧的断点面板BREAKPOINTS，原来勾选了Uncaught Exceptions , 取消勾选即可，谢谢。
